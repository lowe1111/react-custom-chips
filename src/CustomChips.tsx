import React from 'react';

import SearchInput from './SearchInput';
import { ChipData, RemovableChipData } from './chip.interface';
import SampleChip from './SampleChip';
import { ChipsInputContainer, ChipsWrapper } from './chip.styles';


interface Props {
  renderChip: (chipp: RemovableChipData) => JSX.Element;
  renderItem: (selected: boolean, value: ChipData, handleSelect: (val: ChipData) => void) => JSX.Element;
  onChange: (item: ChipData[]) => void;
  inputPlaceholder: string;
  chipsData: ChipData[];
  emptyMessage: string;
  fetchSearchSuggestions?: (value: string) => Promise<ChipData[]>;
  searchIcon?: JSX.Element;
  suggestionList?: ChipData[];
}

interface State {
  chipsData: ChipData[];
}

class CustomChips extends React.Component<Props, State> {
  static defaultProps = {
    searchIcon: <div>Search</div>,
    chipsData: [],
    suggestionList: [],
    renderChip: (value: RemovableChipData) => (<SampleChip key={value.id} value={value} />),
  };

  input: HTMLInputElement | null = null;

  constructor(props: Props) {
    super(props);
    this.state = {
      chipsData: props.chipsData,
    };
  }

  onKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Backspace' && event.currentTarget.value === '') {
      const tmpChipsData = [...this.state.chipsData];
      tmpChipsData.pop();
      this.setState({ chipsData: tmpChipsData });
      this.props.onChange(tmpChipsData);
    }
  }

  addItem = (item: ChipData) => {
    console.log(item)
    const tmpChipsData = [...this.state.chipsData];
    tmpChipsData.push(item);
    this.setState({ chipsData: tmpChipsData });
    console.log(tmpChipsData);
    this.props.onChange(tmpChipsData);
  }

  removeChip = (chip: ChipData) => {
    const tmpChipsData = [...this.state.chipsData];
    const filteredData = tmpChipsData.filter((item) => item.id !== chip.id);
    this.setState({ chipsData: filteredData });
    this.props.onChange(filteredData);
  }

  renderChip = (chip: ChipData) => {
    const tmpChip: RemovableChipData = {
      ...chip,
      onRemove: this.removeChip,
    };
    return this.props.renderChip(tmpChip);
  }

  inputSetting = (input: HTMLInputElement) => {
    this.input = input;
  }

  onKeyDownItem = (event: React.KeyboardEvent<HTMLInputElement>) => {
    this.onKeyPress(event);
  }

  onClickItem = () => {
    if (this.input) {
      this.input.focus();
    }
  }

  render() {
    return (
      <ChipsInputContainer
          onKeyDown={this.onKeyDownItem}
          onClick={this.onClickItem}
      >
        <div>{this.props.searchIcon}</div>
        <ChipsWrapper>
          {this.state.chipsData && this.state.chipsData.map((item) => (
              this.renderChip(item)
          ))}
          <SearchInput
            fetchSearchSuggestions={this.props.fetchSearchSuggestions}
            suggestionList={this.props.suggestionList}
            minLength={1}
            inputClassName="chips-input"
            debounceTimeout={250}
            handleSelectElement={this.addItem}
            renderListItem={this.props.renderItem}
            setInputRef={this.inputSetting}
            inputPlaceholder={this.props.inputPlaceholder}
            emptyMessage={this.props.emptyMessage}
          />
        </ChipsWrapper>
      </ChipsInputContainer>
    );
  }
}
export default CustomChips;
