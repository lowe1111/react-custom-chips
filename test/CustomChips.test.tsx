import React from 'react';
import Enzyme, { mount, render, shallow } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import CustomChips from "../src/CustomChips";

describe('SampleChip', () => {
  beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });

  test('renders correctly', () => {
    const component = shallow(
      <CustomChips

      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  test('shows passed chips', () => {
    const chipsData = [
      { id: '1', name: 'one' },
      { id: '2', name: 'two' },
      { id: '3', name: 'three' },
      { id: '4', name: 'four' },
    ];
    const component = render(
      <CustomChips
        chipsData={chipsData}
      />,
    );
    const chips = component.find('.chip');
    expect(chips.length).toEqual(chipsData.length)
  });
  test('remove one chip works', () => {
    const chipsData = [
      { id: '1', name: 'one' },
      { id: '2', name: 'two' },
      { id: '3', name: 'three' },
      { id: '4', name: 'four' },
    ];
    const component = mount(
      <CustomChips
        chipsData={chipsData}
      />,
    );
    const chip = component.find('.chips_wrapper').find('.chip').first();
    chip.find('.remove').simulate('click');
    expect(component.find('.chips_wrapper').find('.chip').length).toEqual(chipsData.length - 1)
  });
});
