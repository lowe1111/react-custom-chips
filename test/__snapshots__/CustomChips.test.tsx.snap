// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleChip renders correctly 1`] = `
<div
  className="chips_input_container undefined"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <div>
    <SearchIcon />
  </div>
  <div
    className="chips_wrapper"
  >
    <SearchInput
      debounceTimeout={250}
      emptyMessage="empty"
      handleClickOutside={[Function]}
      handleSelectElement={[Function]}
      inputClassName="chips-input"
      inputPlaceholder="Search"
      minLength={1}
      renderListItem={[Function]}
      setInputRef={[Function]}
      suggestionList={Array []}
    />
  </div>
</div>
`;
